Rewrite HTTP server definitions section (#423)
open-telemetry/semantic-conventions: docs/http/http-spans.md

## HTTP server

Read the following section to understand how HTTP server instrumentations are suggested to capture server information.

### HTTP server definitions

An HTTP request can be routed to a specific HTTP application via intermediaries such as reverse proxies.
HTTP requests sent to the same domain name may be handled by multiple applications depending on the port, path, headers, or other parameters.

For example, different versions of the same web-application can run side-by-side as independent applications behind the reverse proxy which routes request to one or another based on the request path.

Instances of different HTTP server applications may run on the same physical host and share the same IP address, but listen to different TCP/UDP ports.

> In order to route request to a specific application, reverse proxies usually modify the [HTTP Host header][Host and authority] replacing the original value provided by the client with an actual proxied server name. This behavior depends on the reverse proxy configuration. In some cases, `Host` header is not used when routing request to a specific application, making it prone to having bogus content.

Reviewer: ```suggestion
In order to route the request to a specific application, reverse proxies usually modify the [HTTP Host header][Host and authority] replacing the original value provided by the client with an actual proxied server name. This behavior depends on the reverse proxy configuration. In some cases, the `Host` header is not used when routing a request to a specific application, making it prone to having bogus content.
```

HTTP server frameworks and their instrumentations have limited knowledge about the HTTP infrastructure and intermediaries that requests go through. In a general case, they can only use HTTP request properties such as request target or headers to populate `server.*` attributes.

#### Setting `server.address` and `server.port` attributes

> In the context of HTTP server, `server.address` and `server.port` attributes capture the original host name and port. The are intended, whenever possible, to be the same on the client and server sides.

Reviewer: ```suggestion
In the context of HTTP server, `server.address` and `server.port` attributes capture the original host name and port. They are intended, whenever possible, to be the same on the client and server sides.
```

## HTTP server

> Read the following section to understand how HTTP server instrumentations should capture server information.

Reviewer 2: If this isn't intended to be normative, please remove `should` for any other word.

e.g.

`Read the following section to understand how HTTP server instrumentations are suggested to capture server information.`

HTTP server frameworks and their instrumentations have a limited knowledge about the HTTP infrastructure and intermediaries request went through. In a general case, they can only use HTTP request properties such as request target or headers to populate `server.*` attributes.

Certain web frameworks may provide additional information about server. For example CGI specification defines `SERVER_NAME` and `SERVER_PORT` ([RFC 3875][rfc-servername]) variables. HTTP servers usually know server names as they are provided by users in the configuration. See for example, the Apache [`ServerName`][ap-sn] or NGINX [`server_name`][nx-sn] directive.

Still, the original logical service name might not be available to the server HTTP instrumentation and instrumentations MAY provide an option to let users set `server.*` attributes.

#### Setting `server.address` and `server.port` attributes

HTTP server instrumentations SHOULD do the best effort when populating `server.address` and `server.port` attributes and determine them by using the first of the following that applies:

* Explicit user configuration (if supported).

> * Server name and port available through web framework or HTTP server configuration (if known to the instrumentation).

Reviewer 3: wondering if this sounds like you should use [ServletRequest#getServerName()](https://docs.oracle.com/javaee/7/api/javax/servlet/ServletRequest.html#getServerName--) and [ServletRequest#getServerName()](https://docs.oracle.com/javaee/7/api/javax/servlet/ServletRequest.html#getServerPort--) (which I think isn't what we want)

Author: Agree. I thought more about it and

- removed explicit config and added a remark that applications can overwrite anything with processors.
- gave the highest priority to the `Forwarded` and friends as they'd match what the client has set
- clarified that framework/web-server info should come from users (and not from `Host` header)

thoughts?

* The [request target](https://www.rfc-editor.org/rfc/rfc9110.html#target.resource) if it's passed in the absolute form.
* The original host which may be passed by the reverse proxy in the `host` identifier of the [`Forwarded`][Forwarded] header or in the [`X-Forwarded-Host`][X-Forwarded-Host] header.
* The `Host` header.

> As a result, `server.*` attributes capture logical host and port name and intended to be the same on the client and server sides. Application developers MAY capture private host information using applicable [resource attributes](/docs/resource/README.md).

Reviewer 3: this might be nice up at the top instead, as it provides the motivation for our definition

## HTTP server

> Read the following section to understand how HTTP server instrumentations capture server information.

Reviewer 3: ```suggestion
Read the following section to understand how HTTP server instrumentations should capture server information.
```

Reviewer 2: should is normative and must be capitalized, FYI

Author: this part is informational and non-normative. There is a normative section ("Setting `server.address` and `server.port` attributes") where normative verbs are capitalized.

Reviewer 2: I know this is awkward, but we should avoid any normative verb in this section of docs.

### HTTP server definitions

This section gives a short summary of some concepts in web server configuration and web app deployment that are relevant to observability.

An HTTP request can be routed to a specific HTTP application via intermediaries such as reverse proxies.
HTTP requests sent to the same domain name may be handled by multiple applications depending on the port, path, headers, or other parameters.

For example, different versions of the same web-application can run side-by-side as independent applications behind the reverse proxy which routes request to one or another based on the request path.

Instances of different HTTP server applications may run on the same physical host and share the same IP address, but listen to different TCP/UDP ports.
In order to route request to a specific application, reverse proxies usually modify the [HTTP Host header][] replacing the original value provided by the client with an actual proxied server name. This behavior depends on the reverse proxy configuration. In some cases, `Host` header is not used when routing request to a specific application, making it prone to having bogus content.

> HTTP server frameworks and their instrumentations have a limited knowledge about the HTTP infrastructure and intermediaries request went through. In a general case, they can only use HTTP request properties such as request target or headers to populate `server.*` attributes.

Reviewer 3: ```suggestion
HTTP server frameworks and their instrumentations have limited knowledge about the HTTP infrastructure and intermediaries that requests go through. In a general case, they can only use HTTP request properties such as request target or headers to populate `server.*` attributes.
```

Still, the original logical server name might not be available to the HTTP server instrumentation or be inaccurate.

#### Setting `server.address` and `server.port` attributes

In the context of HTTP server, `server.address` and `server.port` attributes capture logical host and port name and intended, whenever possible, to be the same on the client and server sides.

HTTP server instrumentations SHOULD do the best effort when populating `server.address` and `server.port` attributes and determine them by using the first of the following that applies:

* The original host which may be passed by the reverse proxy in the `host` identifier of the [`Forwarded`][Forwarded] header or in the [`X-Forwarded-Host`][X-Forwarded-Host] header.
* Server name and port available through web framework or HTTP server properties if they're based on the user configuration. Instrumentations SHOULD NOT use properties that rely on the `Host` header.
* The [request target][request target] if it's passed in the absolute form.

> * The `Host` header.

Reviewer 3: Fallback for HTTP/1.0 where no `Host` header? (`network.local.address/port`?)

Author: Let's figure out `server.address` requirement level first (#424).
If it stays recommended, then I'd prefer to not provide a fallback option.

HTTP 1.0 is [obsolete since 1997](https://en.wikipedia.org/wiki/HTTP#:~:text=Summary%20of%20HTTP%20milestone%20versions) and it would be nice not to add features for it specifically.

Reviewer 3: > Let's figure out `server.address` requirement level first (#424). If it stays recommended, then I'd prefer to not provide a fallback option.

I'm good with keeping `server.address` as (only) recommended, and sent #429 with your proposal for fixing the requirement level on `server.port` instead.

I'm not sure (yet) about not having a fallback.

We've discussed the idea that reconstructing the original URL on the server side is an important use case.

I think users seeing something like `http://123.123.123.123:8080/petclinic` is potentially more useful than not seeing anything.

Author: If address is recommended, then backends should expect that it might be missing. By keeping it recommended, we're also assuming that URL reconstruction is not essential. But it can still be done by falling back (on the consumer) to network attributes.

It also feels like something that can be changed post-stability without breaking based on real-world feedback.

> **Note**: The `Host` header specifies host and port number of the server. The same applies to the `host` identifier of `Forwarded` header, the `X-Forwarded-Host` header, or the authority component in the request target. Instrumentations SHOULD populate both `server.address` and `server.port` attributes by parsing the applicable property.

> 
> The `server.address` SHOULD NOT be set if only IP address is available and capturing name would require a reverse DNS lookup.

Reviewer 3: #413

```suggestion
```
